# Exploration 2

손글씨, 와인, 유방암 데이터셋을 가지고 다양한 모델을 활용하여 분류하고 정확도를 측정하는 project ~!

## 프로젝트 (1) load_digits : 손글씨 분류

### (1) 필요한 모듈 import

In [10]:
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

sklearn.datesets에서 digits(손글씨) 데이터를 import한다

### (2) 데이터 준비

In [3]:
digits=load_digits()
digits.keys()

dict_keys(['data', 'target', 'frame', 'feature_names', 'target_names', 'images', 'DESCR'])

손글씨 데이터셋에는 위와같이 7개의 키들이 들어있음을 확인할 수 있다.   
각각의 데이터들을 확인해보면,

In [4]:
digits_data=digits.data
digits_data #8*8의 픽셀값을 확인 할 수 있음


digits.target_names #label 10개의 target을 확인할 수 있음

(1797, 64)


array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

### (3) 데이터 이해하기

* Feature Data 지정하기   


In [16]:
print(digits_data.shape) #1797 데이터와 64개 features

(1797, 64)


1797개의 데이터셋이 있으며 각각의 데이터는 64개의 특징을 가진다.

* Label Date 지정하기

In [5]:
digits_label=digits.target
digits_label

array([0, 1, 2, ..., 8, 9, 8])

손글씨의 경우 레이블은 0~9까지 있다.

* Target Names 출력해 보기

In [6]:
digits.target_names

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

* 데이터 Describe 해 보기

In [7]:
print(digits.DESCR) #데이터셋에 대한 설명!

.. _digits_dataset:

Optical recognition of handwritten digits dataset
--------------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 1797
    :Number of Attributes: 64
    :Attribute Information: 8x8 image of integer pixels in the range 0..16.
    :Missing Attribute Values: None
    :Creator: E. Alpaydin (alpaydin '@' boun.edu.tr)
    :Date: July; 1998

This is a copy of the test set of the UCI ML hand-written digits datasets
https://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits

The data set contains images of hand-written digits: 10 classes where
each class refers to a digit.

Preprocessing programs made available by NIST were used to extract
normalized bitmaps of handwritten digits from a preprinted form. From a
total of 43 people, 30 contributed to the training set and different 13
to the test set. 32x32 bitmaps are divided into nonoverlapping blocks of
4x4 and the number of on pixels are counted in each blo

### (4) train, test 데이터 분리

In [19]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(digits_data, 
                                                    digits_label, 
                                                    test_size=0.2, #노드와 동일하게 20%를 테스트셋으로 분류하였다.
                                                    random_state=7)

print('X_train 개수: ', len(X_train), ', X_test 개수: ', len(X_test))

X_train 개수:  1437 , X_test 개수:  360


### (5) 다양한 모델로 학습시켜보기

* Decision Tree 사용해 보기

In [21]:
#Decision Tree 모델 불러오기

from sklearn.tree import DecisionTreeClassifier

decision_tree = DecisionTreeClassifier(random_state=32)
decision_tree.fit(X_train,y_train) #DT 모델 학습!
y_pred = decision_tree.predict(X_test)
y_pred

from sklearn.metrics import accuracy_score #sklearn.metrics패키지를 활용하여 정확도를 확인

accuracy = accuracy_score(y_test, y_pred)
accuracy # 약 85% 정도의 정확도를 보인다  

0.8555555555555555

* Random Forest 사용해 보기

In [22]:
# Randome Forest 패키지 import!

from sklearn.ensemble import RandomForestClassifier

# train 시키기

random_forest = RandomForestClassifier(random_state=32)
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)

# 정확도 계산!

random_forest = DecisionTreeClassifier(random_state=32)
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       0.81      0.81      0.81        42
           2       0.79      0.82      0.80        40
           3       0.79      0.91      0.85        34
           4       0.83      0.95      0.89        37
           5       0.90      0.96      0.93        28
           6       0.84      0.93      0.88        28
           7       0.96      0.82      0.89        33
           8       0.88      0.65      0.75        43
           9       0.78      0.78      0.78        32

    accuracy                           0.86       360
   macro avg       0.86      0.86      0.86       360
weighted avg       0.86      0.86      0.85       360



랜덤트리의 경우 의사결정나무 보다 약간 더 높은 86%의 정확도를 보인다.

* SVM 사용해 보기

In [23]:
#SVM(Support Vector Machine) 모델 불러오기

from sklearn import svm
svm_model = svm.SVC()

svm_model.fit(X_train, y_train) #SVM 모델로 트래인 시킨 후

svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       0.95      1.00      0.98        42
           2       1.00      1.00      1.00        40
           3       1.00      1.00      1.00        34
           4       1.00      1.00      1.00        37
           5       0.93      1.00      0.97        28
           6       1.00      1.00      1.00        28
           7       1.00      1.00      1.00        33
           8       1.00      0.93      0.96        43
           9       1.00      0.97      0.98        32

    accuracy                           0.99       360
   macro avg       0.99      0.99      0.99       360
weighted avg       0.99      0.99      0.99       360



SVM은 거의 완벽에 가까운 평균 99%의 정확도를 보인다. 

* SGD Classifier 사용해 보기

In [24]:
# SGD(Stochastic Gradient Descent) Classifier 패키지를 불러온다

from sklearn.linear_model import SGDClassifier
sgd_model = SGDClassifier()

#SGD 모델로 다시 학습을 시킨 후,
sgd_model.fit(X_train, y_train)

#결과값을 출력한다

sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       0.86      0.90      0.88        42
           2       0.98      1.00      0.99        40
           3       0.87      0.97      0.92        34
           4       0.97      0.97      0.97        37
           5       0.96      0.96      0.96        28
           6       0.96      0.93      0.95        28
           7       0.89      0.97      0.93        33
           8       0.94      0.79      0.86        43
           9       0.93      0.88      0.90        32

    accuracy                           0.94       360
   macro avg       0.94      0.94      0.94       360
weighted avg       0.94      0.94      0.94       360



SGD 분류기는 약 94퍼센트의 정확도가 나왔다.

* Logistic Regression 사용해 보기 

In [26]:
#Logistic Regression 모델을 import한다


from sklearn.linear_model import LogisticRegression
logistic_model = LogisticRegression()

logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)

logistic_model = LogisticRegression(solver='lbfgs', max_iter=10000).fit(X_train, y_train)

print(classification_report(y_test, y_pred))

/home/aiffel/anaconda3/envs/datascience/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       0.95      0.95      0.95        42
           2       0.98      1.00      0.99        40
           3       0.94      0.97      0.96        34
           4       0.97      1.00      0.99        37
           5       0.82      0.96      0.89        28
           6       1.00      0.96      0.98        28
           7       0.97      0.97      0.97        33
           8       0.92      0.81      0.86        43
           9       0.97      0.91      0.94        32

    accuracy                           0.95       360
   macro avg       0.95      0.95      0.95       360
weighted avg       0.95      0.95      0.95       360



Logistic regression은 평균 약 95 퍼센트의 정확도를 보인다.

### (6) 모델을 평가해 보기

같은 데이값과 테스트셋을 가지고 모델마다 다른 정확도를 보인다.   
손글씨 분류기같은 경우는 레이블이 10개나 나눠져 있었음에도 랜덤트리 모델에서 99%의 정확도가 나왔다.   
와인과 유방암 분류에서는 train set과 test set의 분류를 더 많이 한 후 정확도를 비교해보겠다.

# 프로젝트 (2) load_wine : 와인을 분류

### (1) 필요한 모듈 import

In [27]:
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report


```sklearn.datesets```에서 wine(와인) 데이인셋을 ```import```한다.

### (2) 데이터 준비

In [28]:
wine = load_wine()
wine.keys()

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names'])

```load_wine```에는 위와 같이 6개의 키가 들어있다.   
각 키에 대한 정보는 아래 **(3)데이터 이해하기**에서 확인해보겠다.

### (3) 데이터 이해하기

    1) Feature Data 지정하기

In [29]:
wine_data = wine.data
print(wine_data.shape)

(178, 13)


와인의 데이터셋에는 178개의 와인데이와 각 13개의 feature가 들어가 있다.

    2) Label Data 지정하기

In [30]:
wine_label = wine.target
wine_label

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2])

와인 레이블에는 0, 1, 2 세개의 class로 나뉘는 것을 확인할 수 있다.

    3) Target Nmaes 출력

In [31]:
wine.target_names

array(['class_0', 'class_1', 'class_2'], dtype='<U7')

레이블의 구분은 Class_0, Class_1, Class_2 로 나와있음

    4) 데이터 Describe

In [32]:
print(wine.DESCR)

.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

    :Number of Instances: 178 (50 in each of three classes)
    :Number of Attributes: 13 numeric, predictive attributes and the class
    :Attribute Information:
 		- Alcohol
 		- Malic acid
 		- Ash
		- Alcalinity of ash  
 		- Magnesium
		- Total phenols
 		- Flavanoids
 		- Nonflavanoid phenols
 		- Proanthocyanins
		- Color intensity
 		- Hue
 		- OD280/OD315 of diluted wines
 		- Proline

    - class:
            - class_0
            - class_1
            - class_2
		
    :Summary Statistics:
    
    ============================= ==== ===== ======= =====
                                   Min   Max   Mean     SD
    ============================= ==== ===== ======= =====
    Alcohol:                      11.0  14.8    13.0   0.8
    Malic Acid:                   0.74  5.80    2.34  1.12
    Ash:                          1.36  3.23    2.36  0.27
    Alcalinity of Ash:            1

```load_wine()``` 데이터셋에 대한 정보를 확인할 수 있음!

### (4) train, test 데이터 분리

In [33]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(wine_data, 
                                                    wine_label, 
                                                    test_size=0.2, #노드와 동일하게 20%를 테스트셋으로 분류하였다.
                                                    random_state=7)

print('X_train 개수: ', len(X_train), ', X_test 개수: ', len(X_test))

X_train 개수:  142 , X_test 개수:  36


데이터셋의 20%를 테스트셋으로 분류한 결과, train set 142개, test set 36개로 분류되었다.

### (5) 다양한 모델로 학습시켜보기

    1) Decision Tree

In [34]:
#Decision Tree 모델 불러오기

from sklearn.tree import DecisionTreeClassifier

decision_tree = DecisionTreeClassifier(random_state=32)
decision_tree.fit(X_train, y_train) #DT 모델 학습!

print('X_train 개수: ', len(X_train), ', X_test 개수: ', len(X_test))

print("Wine dataset을 의사결정나무 모델로 학습시켰다.")

X_train 개수:  142 , X_test 개수:  36
Wine dataset을 의사결정나무 모델로 학습시켰다.


이제는 ```sklearn.metrics``` 패키지를 이용하여 정확도를 계산해본다.

In [55]:
from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test, y_pred)
accuracy

0.8671328671328671

 86%의 정확도가 나왔다@@@

    2) Random Forest

In [ ]:
# Randome Forest 패키지 import!

from sklearn.ensemble import RandomForestClassifier

# train 시키기

random_forest = RandomForestClassifier(random_state=32)
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)

# 정확도 계산!

random_forest = DecisionTreeClassifier(random_state=32)
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)

print(classification_report(y_test, y_pred))

random tree의 경우에됴 평균 94%에 정확도가 나왔다.   
앞서 손글씨 분류기와 동일하게 random tree가 decision tree보다 높은 정확도를 보이고 있다!

    3) SVM(Support Vector Machine)

In [51]:
from sklearn import svm
svm_model = svm.SVC()

svm_model.fit(X_train, y_train) #SVM 모델로 트래인 시킨 후

y_pred = svm_model.predict(X_test) #결과값을 예측하여

print(classification_report(y_test, y_pred)) #프린트한다.

              precision    recall  f1-score   support

           0       1.00      0.78      0.88        45
           1       0.91      1.00      0.95        98

    accuracy                           0.93       143
   macro avg       0.95      0.89      0.91       143
weighted avg       0.94      0.93      0.93       143



SVM 모델의 경우 다른 두 모델보다는 정확도가 약간 떨어지는걸 볼 수 있다.    
평균 정확도 61%

    4) SGD(Stochastic Gradient Descent) Classifier 

In [54]:
from sklearn.linear_model import SGDClassifier
sgd_model = SGDClassifier()

#SGD 모델로 다시 학습을 시킨 후,
sgd_model.fit(X_train, y_train)

#결과값을 출력한다

sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.75      0.87      0.80        45
           1       0.93      0.87      0.90        98

    accuracy                           0.87       143
   macro avg       0.84      0.87      0.85       143
weighted avg       0.88      0.87      0.87       143



SGD는 69퍼센트의 정확도를 보인다. SVM 보다는 소폭 상승하였다.

    5) Logistic Regression

In [37]:
from sklearn.linear_model import LogisticRegression
logistic_model = LogisticRegression()

logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)

logistic_model = LogisticRegression(solver='lbfgs', max_iter=10000).fit(X_train, y_train)

print(classification_report(y_test, y_pred)).

/home/aiffel/anaconda3/envs/datascience/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


              precision    recall  f1-score   support

           0       1.00      0.86      0.92         7
           1       0.94      1.00      0.97        17
           2       1.00      1.00      1.00        12

    accuracy                           0.97        36
   macro avg       0.98      0.95      0.96        36
weighted avg       0.97      0.97      0.97        36



logistic regression은 97퍼센트의 정확도를 보여준다!

### (6) 모델을 평가해 보기

와인 데이터셋에서는 Logistic regressioin이 가장 높은 정확도를 보인다.   
아직 각 모델들이 어떻게 다른 값을 나타내가 되는지 모델들에 대한 깊은 이해는 하지 못했다.   
시간을 가지고 모델들에 대한 충분한 이해가 필요하다.

## 프로젝트 (3) load_breast_cancer : 유방암 여부를 진단

### (1) 필요한 모듈 import하기

In [56]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

```load_breast_cancer```데이터셋을 ```import``` 합니다.

### (2) 데이터 준비

In [57]:
cancer = load_breast_cancer()
cancer.keys()

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename'])

유방암 데이터셋의 키 값들도 **(1) digits**, **(2) wine** 데이터셋과 동일하게 6개의 키값이 들어 있습니다.

### (3) 데이터 이해하기

    1) Feature Data 지정

In [58]:
cancer_data = cancer.data
print(cancer_data.shape)

(569, 30)


유방암 데이터셋에는 569개의 암 데이터와 각 30개의 features가 들어가 있습니다.

    2) Label Data 지정하기

In [59]:
cancer_label = cancer.target
cancer_label

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
       1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0,
       0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1,
       1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0,
       0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0,
       1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0,

유방암 데이터셋의 target은 0과 1로 이루어져 있으며 무작위로 섞여있는 듯한 느낌이 듭니다.   
이어서 바로 target name들을 확인해볼게요. 어떤걸로 구성되어 있을지 너무 궁금하네요.

    3) Target Nmaes 출력

In [60]:
cancer.target_names

array(['malignant', 'benign'], dtype='<U9')

악성종양과 양성종양 그리고 U9이라는 것으로 이루어져 있다. 아직 U9이 label인지 아닌지는 잘 모르겠다.

In [61]:
import pandas as pd
cancer_df = pd.DataFrame(data=cancer_data, columns=cancer.feature_names)
cancer_df

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
0             0.27760         0.30010              0.14710         0.2419   
1             0.07864         0.08690              0.07017         0.1812   
2             0.15990         0.19740              0.12790         0.2069   
3             0.28390         0.24140              0.10520         0.2597   
4             0.13280         0.19800              0.10430         0.1809   
..                ...             ...                  ...            ...   
564           0.11590         0.24390              0.13890         0.1726   
565           0.10340         0.14400              0.09791         0.1752   
566           0.10230         0.09251              0.05302         0.1590   
567           0.27700         0.35140              0.15200         0.2397   
568           0.04362         0.00000              0.00000         0.1587   

     mean fractal dimension  ...  worst radius  worst texture  \
0                   0.07871  ...        25.380          17.33   
1                   0.05667  ...        24.990          23.41   
2                   0.05999  ...        23.570          25.53   
3                   0.09744  ...        14.910          26.50   
4                   0.05883  ...        22.540          16.67   
..                      ...  ...           ...            ...   
564                 0.05623  ...        25.450          26.40   
565                 0.05533  ...        23.690          38.25   
566                 0.05648  ...        18.980          34.12   
567                 0.07016  ...        25.740          39.42   
568                 0.05884  ...         9.456          30.37   

     worst perimeter  worst area  worst smoothness  worst compactness  \
0             184.60      2019.0           0.16220            0.66560   
1             158.80      1956.0           0.12380            0.18660   
2             152.50      1709.0           0.14440            0.42450   
3              98.87       567.7           0.20980            0.86630   
4             152.20      1575.0           0.13740            0.20500   
..               ...         ...               ...                ...   
564           166.10      2027.0           0.14100            0.21130   
565           155.00      1731.0           0.11660            0.19220   
566           126.70      1124.0           0.11390            0.30940   
567           184.60      1821.0           0.16500            0.86810   
568            59.16       268.6           0.08996            0.06444   

     worst concavity  worst concave points  worst symmetry  \
0             0.7119                0.2654          0.4601   
1             0.2416                0.1860          0.2750   
2             0.4504                0.2430          0.3613   
3             0.6869                0.2575          0.6638   
4             0.4000                0.1625          0.2364   
..               ...                   ...             ...   
564           0.4107                0.2216         

여기에 label을 입혀서 다시 한번 확인해보겠다. 그럼 u9에 대해서 알 수도 있을것 같다.

In [62]:
cancer_df["label"]=cancer.target
cancer_df

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
0             0.27760         0.30010              0.14710         0.2419   
1             0.07864         0.08690              0.07017         0.1812   
2             0.15990         0.19740              0.12790         0.2069   
3             0.28390         0.24140              0.10520         0.2597   
4             0.13280         0.19800              0.10430         0.1809   
..                ...             ...                  ...            ...   
564           0.11590         0.24390              0.13890         0.1726   
565           0.10340         0.14400              0.09791         0.1752   
566           0.10230         0.09251              0.05302         0.1590   
567           0.27700         0.35140              0.15200         0.2397   
568           0.04362         0.00000              0.00000         0.1587   

     mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                   0.07871  ...          17.33           184.60      2019.0   
1                   0.05667  ...          23.41           158.80      1956.0   
2                   0.05999  ...          25.53           152.50      1709.0   
3                   0.09744  ...          26.50            98.87       567.7   
4                   0.05883  ...          16.67           152.20      1575.0   
..                      ...  ...            ...              ...         ...   
564                 0.05623  ...          26.40           166.10      2027.0   
565                 0.05533  ...          38.25           155.00      1731.0   
566                 0.05648  ...          34.12           126.70      1124.0   
567                 0.07016  ...          39.42           184.60      1821.0   
568                 0.05884  ...          30.37            59.16       268.6   

     worst smoothness  worst compactness  worst concavity  \
0             0.16220            0.66560           0.7119   
1             0.12380            0.18660           0.2416   
2             0.14440            0.42450           0.4504   
3             0.20980            0.86630           0.6869   
4             0.13740            0.20500           0.4000   
..                ...                ...              ...   
564           0.14100            0.21130           0.4107   
565           0.11660            0.19220           0.3215   
566           0.11390            0.30940           0.3403   
567           0.16500            0.86810           0.9387   
568           0.08996            0.06444           0.0000   

     worst concave points  worst symmetry  worst fractal dimension  label  
0                  0.2654          0.4601                  0.11890      0  
1                  0.1860          0.2750                  0.08902      0  
2                  0.2430          0.3613                  0.08758      0  
3                  0.2575          0.6638                  0.17300      0  
4                  0.1625          0.2364                  0.07678   

확인을 해보니 label은 1, 0으로만 나뉜다. 즉, 'malignant', 'benign'으로만 확인하면 될것 같다.

    4) 데이터 Describe

In [63]:
print(cancer.DESCR)

.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 569

    :Number of Attributes: 30 numeric, predictive attributes and the class

    :Attribute Information:
        - radius (mean of distances from center to points on the perimeter)
        - texture (standard deviation of gray-scale values)
        - perimeter
        - area
        - smoothness (local variation in radius lengths)
        - compactness (perimeter^2 / area - 1.0)
        - concavity (severity of concave portions of the contour)
        - concave points (number of concave portions of the contour)
        - symmetry
        - fractal dimension ("coastline approximation" - 1)

        The mean, standard error, and "worst" or largest (mean of the three
        worst/largest values) of these features were computed for each image,
        resulting in 30 features.  For instance, field 0 is Mean Radi

### (4) train, test 데이터 분리

In [64]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(cancer_data, 
                                                    cancer_label, 
                                                    test_size=0.25, #이번에는 25%를 테스트셋으로 분류해보겠다.
                                                    random_state=7)

print('X_train 개수: ', len(X_train), ', X_test 개수: ', len(X_test))

X_train 개수:  426 , X_test 개수:  143


75 : 25의 비율로 train set과  test set을 분류하였다.   
426개와 143개로 나뉜 것을 확인할 수 있다.

### (5) 다양한 모델로 학습시켜보기

모델 마다의 값이 너무 궁금하다. 학습을 시킨 후 테스트 결과를 보겠다.

    1) Decision Tree

In [65]:
#Decision Tree 모델 불러오기

from sklearn.tree import DecisionTreeClassifier

decision_tree = DecisionTreeClassifier(random_state=32)
decision_tree.fit(X_train, y_train) #DT 모델 학습!



DecisionTreeClassifier(random_state=32)

In [66]:
print('X_train 개수: ', len(X_train), ', X_test 개수: ', len(X_test))

print("cancer dataset을 의사결정나무 모델로 학습시켰다.")

X_train 개수:  426 , X_test 개수:  143
cancer dataset을 의사결정나무 모델로 학습시켰다.


이제는 ```sklearn.metrics``` 패키지를 이용하여 정확도를 계산해본다.

In [67]:
from sklearn.metrics import accuracy_score

In [68]:
accuracy1 = accuracy_score(y_test, y_pred)
accuracy1

0.8671328671328671

원래는 알수없는 에러메세지와 함께 안됐었는데 이제는 된다. 뭐가 문젠지는 모르겠다.   
Decision tree는 모든 데이터셋에서 나쁘지 않은 정확도를 보이고 있다.

    2) Random Forest

In [69]:
# Randome Forest 패키지 import!

from sklearn.ensemble import RandomForestClassifier

# train 시키기

random_forest = RandomForestClassifier(random_state=32)
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)

# 정확도 계산!

random_forest = DecisionTreeClassifier(random_state=32)
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.95      0.78      0.85        45
           1       0.91      0.98      0.94        98

    accuracy                           0.92       143
   macro avg       0.93      0.88      0.90       143
weighted avg       0.92      0.92      0.91       143



random tree의 경우에됴 평균 92%의 정확도가 나왔다.

    3) SVM(Support Vector Machine)

In [70]:
from sklearn import svm
svm_model = svm.SVC()

svm_model.fit(X_train, y_train) #SVM 모델로 트래인 시킨 후

y_pred = svm_model.predict(X_test) #결과값을 예측하여

print(classification_report(y_test, y_pred)) #프린트한다.

              precision    recall  f1-score   support

           0       1.00      0.78      0.88        45
           1       0.91      1.00      0.95        98

    accuracy                           0.93       143
   macro avg       0.95      0.89      0.91       143
weighted avg       0.94      0.93      0.93       143



SVM 모델은 **(2)wine**의 경우에 정확도가 많이 떨어졌었는데 유방암 데이터셋에는 다시 높은 정확도를 보이고 있다.

    4) SGD(Stochastic Gradient Descent) Classifier 
   

In [71]:
from sklearn.linear_model import SGDClassifier
sgd_model = SGDClassifier()

#SGD 모델로 다시 학습을 시킨 후,
sgd_model.fit(X_train, y_train)

#결과값을 출력한다

sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.94      0.76      0.84        45
           1       0.90      0.98      0.94        98

    accuracy                           0.91       143
   macro avg       0.92      0.87      0.89       143
weighted avg       0.91      0.91      0.91       143



SGD 모델은 **(2)wine**의 경우에 SVM 모델보다 높은 정확도를 보여줬었는데 유방암 데이터셋에는 더 낮은 정확도를 보이고 있다.

    5) Logistic Regression

In [72]:
from sklearn.linear_model import LogisticRegression
logistic_model = LogisticRegression()

logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)

logistic_model = LogisticRegression(solver='lbfgs', max_iter=10000).fit(X_train, y_train)

print(classification_report(y_test, y_pred))

/home/aiffel/anaconda3/envs/datascience/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


              precision    recall  f1-score   support

           0       0.97      0.84      0.90        45
           1       0.93      0.99      0.96        98

    accuracy                           0.94       143
   macro avg       0.95      0.92      0.93       143
weighted avg       0.95      0.94      0.94       143



logistic regression은 94퍼센트의 정확도가 나온다.

### (6) 모델을 평가해 보기

모든 모델들을 전부 활용하여 세 데이터셋을 비교한 결과, Logistic regression에서 가장 높은 정확도를 보였다.   
특이하게도 유방암 데이터에서는 다른 데이터셋들과 다르게 SVM 모델이 꽤 높은 정확도를 보여줬다. 이유가 너무 궁금하다.
앞 평가에서 말했듯이, 각 모델들이 어떠한 차이를 가지고 왜 이러한 정확도와 결과값을 가지는지 이해하기에는   
개인적인 역량과 시간이 부족했다.   
Aiffel 과정을 공부하며 각 모델들에 대한 공부를 해야겠다.

### 회고:    

**1. 이번 프로젝트에서 어려웠던 점**   
각 분류 모델들에 대한 차이를 이해하기가 어려웠다.   
모델들에 대한 충분한 이해가 이루어지지 않아 정확도를 도출하는 데에도 왜 그런 결과가 나오는지 이해할 수가 없어서 아쉬웠다..   
시간을 가지고 공부를 하고 싶었지만 아직 다른 노드를 할게 많고, 시간이 부족해서 하지 못했다ㅠㅠ   
   

**2. 프로젝트를 진행하면서 알아낸 점 혹은 아직 모호한 점.**   
모델에 대한 이해 뿐만 아니라 코드에 대한 이해도 아직 부족하다고 생각한다. 개인적인 역량 부족이다.   
코드를 이해하지 못하니 노드를 따라가는데 어려움과 아쉬움이 많이 있었다.   
그래서 파이썬순위로 공부하고 있다.   
   

**3. 루브릭 평가 지표를 맞추기 위해 시도한 것들.**   
개인적으로는 이번 프로젝트를 통해 많은걸 배웠다.   
```.key```값을 통해 데이터셋이 가지고 있는 키값들을 불러오고 각각이 어떤걸 의미하는지 충분히 이해하였다.   
또 여러가지 모델을 활용하여 테스트 및 정확도를 측정하는데 문제가 없었고,
정확도를 기준으로 모델을 선정하였다.
   
   
**4. 자기 다짐**   
개인적인 역량을 더 많이 키워야겠다는 생각이 든다.   
Aiffel은 단거리 달리기가 아니라 장거리 마라톤과 같이 멀리 보고 천천히 오래 공부해야한다. 하지만 나는 단거리 달리기로 다른 사람들과 격차를 좁히고,   
다른 사람들과 비스무리한 수준이 되기 위해서 초반에 단거리 달리기 공부를 할 필요가 있다고 느꼈다.
   
    
**5. 기타**   
여러가지 데이터와 활용방법에 대해 충분히 이해할 수 있어서 좋았다.